### PR TITLE
Add testing section and update documentation to latest state of jsp

### PR DESCRIPTION
**Pull request by maroroman**
Updated the documentation based on API improvements and in anticipation of new UI/API features. Also added a testing section describing how to deploy changes to JSP and how to deploy changes that also include changes to jupyterhub-odh

### CODE REVIEW

File: docs/images.md [hunk 1]
```diff
@@ -13,4 +13,16 @@ You can list images in your cluster by running
...
+For more information about creating a custom image to use in the UI check the [Adding custom image documentation](http://opendatahub.io/docs/administration/installation-customization/add-custom-image.html)
+
+## Annotations
+To provide more information about the image to the user, or sort the images, it is possible to add the following annotations to the image:
+
+- `opendatahub.io/default-image`: `bool`, sets whether the image should be used as default in the Spawner UI
+- `opendatahub.io/notebook-image-desc`: Image description
+- `opendatahub.io/notebook-image-name`: Human readable name
+- `opendatahub.io/notebook-image-url`: Url of image
+- `opendatahub.io/image-order`: Sets the order in which images appear within UI, defaults to last place.
```
Comment (vpavlin) [inline docs/images.md:25]: This changed to `notebook-image-order`

File: docs/images.md [hunk 2]
```diff
@@ -13,4 +13,16 @@ You can list images in your cluster by running
...
+For more information about creating a custom image to use in the UI check the [Adding custom image documentation](http://opendatahub.io/docs/administration/installation-customization/add-custom-image.html)
+
+## Annotations
+To provide more information about the image to the user, or sort the images, it is possible to add the following annotations to the image:
+
+- `opendatahub.io/default-image`: `bool`, sets whether the image should be used as default in the Spawner UI
+- `opendatahub.io/notebook-image-desc`: Image description
+- `opendatahub.io/notebook-image-name`: Human readable name
+- `opendatahub.io/notebook-image-url`: Url of image
+- `opendatahub.io/notebook-image-order`: Sets the order in which images appear within UI, defaults to last place.
```
Comment (crobby) [inline docs/images.md:25]: "defaults to last place" is unclear to me.  What does that mean?

File: docs/testing.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+3. Push the changes in the JSP library
+4. Modify the `jupyterhub-singleuser-profiles` value in the jupyterhub-odh [Pipfile](https://github.com/opendatahub-io/jupyterhub-odh/blob/master/Pipfile) to ``` {ref = "branch-name",git = "https://github.com/<username>/jupyterhub-singleuser-profiles.git"} ``` where the `git` value should point to your fork
+5. Run the `pipenv lock` command. Save all changes and push them to the jupyterhub-odh repo
+6. Access the `jupyterhub-img` `BuildConfig` on your cluster. 
+7. Look for the `uri` value under `source`, Change it to your fork adress. Then add a `ref` value under `uri`, which will contain the name of your branch.
+8. Start the build. When the build finishes, redeploy the jupyterhub pod.
+
+### Notes
+
+- If you are missing the `jupyterhub-img` `BuildConfig`,
```
[the remark this answers is not in the document]
Comment (crobby) [reply docs/testing.md:35]: Yes, this seems like it is incomplete

**Review by crobby — COMMENTED** [review]
Just a couple of questions, but looks good in general

File: docs/api.md [hunk 4]
```diff
@@ -2,17 +2,14 @@
 
 The new API allows gathering data which is shown in the UI.
 The capabilities of the API are:
-  - User ConfigMaps (`GET`, `POST`), It is possible to request the json of the User ConfigMap, and send a body of data to add (or replace). This includes selections, enviroment variables and gpu number.
-  - Sizes (`GET`), Sizes can be requested, either as an array of names (used in the UI) or as a pure json body (using the `pure_json=true` query)
-  - Images (`GET`), Returns an array of images (which have the proper tag). Note: Due to some limitation it is not possible to gather a json with all of the image information
+  - User ConfigMaps (`GET`, `POST`), It is possible to request the json of the User `ConfigMap`, and send a body of data to add (or replace). This includes selections, enviroment variables and gpu number.
+  - UI Configuration (`GET`), If UI Configuration is set in the main `ConfigMap`, the API is able to return information neccessary to configure the visual style and various elements of the UI. More information about this section can be found [here](./ui_configuration.md)
```
Comment (vpavlin) [inline docs/api.md:6]: Not only in main...any JSP configmap would do

File: docs/testing.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mroman-redhat) [reply docs/testing.md:35]: There are actually more than one ways to solve the problem of missing that buildconfig, This happens to basically anyone who is not developing JSP. My way of solving this is using an opendatahub kfdef and applying it to the cluster with kfctl. The problem is which kfdef should be posted here, and whether I should post some link to vague documentation on how to use kfctl or write up exact instructions here.

**Review by crobby — APPROVED** [review]

File: docs/ui_configuration.md [hunk 5]
```diff
@@ -42,61 +42,75 @@ Similarly, if you wanted to make certain images, (ex. minimal-notebook:v0.1 and
...
+- name: {string}
+    variables:
     - name: {string}
-      variables:
-      - name: {string}
-        type: {"password" || "text"}
-      - name: {string}
-        type: {"password" || "text"}
-    ```
+    type: {"password" || "text"}
```
Comment (vpavlin) [inline docs/ui_configuration.md:113]: `type` lines misaligned - see https://github.com/opendatahub-io/jupyterhub-singleuser-profiles/blob/9c3434421b08837a46c2731581af33a92f24acac/docs/ui_configuration.md

**Comment by LaVLaS** [comment]
@maroroman Can we add an description and example for a custom singleuser-profile?  We previously had this [example](https://gitlab.com/opendatahub/opendatahub-operator/-/tree/master/roles/aicoe-jupyterhub#additional-information) in the old odh gitlab org which should still be valid for the current code base

**Comment by maroroman** [comment]
> @maroroman Can we add an description and example for a custom singleuser-profile? We previously had this [example](https://gitlab.com/opendatahub/opendatahub-operator/-/tree/master/roles/aicoe-jupyterhub#additional-information) in the old odh gitlab org which should still be valid for the current code base

Can this be used? Does there need to be any additional information? https://github.com/opendatahub-io/jupyterhub-singleuser-profiles/blob/297a51d697aa14ccb0695d9278b368852669ca68/docs/configuration.md

**Comment by vpavlin** [comment]
lgtm

**Pull request merged**